collector/database/classes/collector.php

```diff
@@ -39,19 +45,41 @@ public function record_metric(metric_item $item) {
         );
     }
 
+    /**
+     * Deletes every metric from cltr table for a given metric name.
+     *
+     * @param string $metricname Metric name to remove.
+     */
+    public function delete_metrics(string $metricname) {
+        global $DB;
+
+        $DB->delete_records(lib::TABLE, ['name' => $metricname]);
+    }
+
     /**
      * Returns stored metrics for the collector.
      *
-     * @param mixed $metricnames The metrics to be retrieved. Either a single string, or an
+     * @param  mixed  $metricnames The metrics to be retrieved. Either a single string, or an
      *         array of strings. If empty, then all available metrics will be retrieved.
-     * @param int | false $starting The earliest timestamp to retrieve.
+     * @param  object $metricsclass Metric class being used.
+     * @param  int | false $starting The earliest timestamp to retrieve.
+     * @param  array Array of data if metric is backfilled.
      * @return array
      * @throws \coding_exception
      * @throws \dml_exception
      */
-    public function get_metrics($metricnames = null, $since = false): array {
+    public function get_metrics($metricnames = null, $metricclass = null, $since = false, $data = []): array {
         global $DB;
         $starting = '';
+
+        if (!is_null($metricclass)) {
+            if ($metricclass->is_backfillable()) {
+                $backfilleddata = $metricclass->get_backfilled_metric($data);
```

Comment:
Is this the best place for it?
This function (which I now realise is poorly named. It should be called 'get_saved_metric_data', or something similar), is about retrieving saved metric data, not about generating them. Perhaps this should be done where metric data is originally generated.

This function is unique to cltr_database\collector. Other collectors may have the ability to record backfill data, but not have the ability to retrieve it (like AWS).

Reply:
I have split this function with others - as stated in #67, I also added a declaration whether or not a collector supports metric recording - should there also be a declaration as whether they can be retrieved or not ?